trillian/migrillian/core/controller.go

```diff
@@ -30,9 +31,36 @@ import (
 
 	"github.com/google/trillian/merkle"
 	_ "github.com/google/trillian/merkle/rfc6962" // Register hasher.
+	"github.com/google/trillian/monitoring"
 	"github.com/google/trillian/types"
 )
 
+// Metrics holds Controller metrics keyed by Tree ID.
+type Metrics struct {
+	masterRuns     monitoring.Counter
+	masterCancels  monitoring.Counter
+	isMaster       monitoring.Gauge
+	entriesFetched monitoring.Counter
+	entriesSeen    monitoring.Counter
+	entriesStored  monitoring.Counter
+	// TODO(pavelkalinnikov): Add latency histograms, latest STH, tree size, etc.
```

Comment:
Our [normal model](https://github.com/google/certificate-transparency-go/blob/master/trillian/ctfe/handlers.go#L260) is to have the metrics as a collection of global variables and have a `sync.Once` that creates them (maybe in `NewController`?).

Reply:
This model effectively assumes that `MetricFactory` is also global (even though it's passed around as a field / local var). In this case we should probably have a global var for factory next to the `once`, to make this assumption explicit.

In this model though I can't create 2 independent instances (say, with 2 different path [prefixes](https://github.com/google/trillian/blob/8515c71db542509da14a7e13139fa16c9e785a0e/monitoring/prometheus/metrics.go#L31)). My current code is more flexible, as it allows `main` to create 2 factories with different prefixes, and pass different `Metrics` structs to `Controllers`. (not that I need this capability at the moment, but I like it being more generic)

Reply:
Yes, it does effectively assume that `MetricFactory` is global and that there is never any need to have different `MetricFactory` instances in the same process.  This reflects the original aim of the `MetricFactory` interface, which is to allow injection of entirely different implementations for different environments.

The `prometheus.MetricFactory.Prefix` field is only used for testing AFAIK; actually having distinct prefixes would cause duplication problems for all of the downstream code that monitors, aggregates and alerts on the metrics.

As for being more generic, I suspect YAGNI :-)

Reply:
Yeah, probably having 2 factories is YAGNI.

I still believe there is no immediate necessity to have globals, because `main` can just create the metrics (effectively achieving the `once` semantic by definition) and pass them to the `Controller`.

Passing the factory in to the `Controller` would be necessary if the latter had to recursively pass it to other packages/susbsystems. But since it's a "leaf" package at the moment, passing the factory in it is itself a potential YAGNI.

Reply:
`main` creating the metrics means that this needs to be done in both open-source and Google-internal versions.  

The pattern used in all the other places that have metrics are used is to inject a `MetricFactory`; there is no benefit to being inconsistent here, so please don't be.

Reply:
Sigh. Done. PTAL.

Reply:
Sigh indeed.